### PR TITLE
{model, runner, internal/session}: add session multimodal externalization

### PR DESCRIPTION
**Pull request by hr-chang**
## Summary

- Add `model.ContentRef` as a unified internal reference for externalized multimodal content in `model.ContentPart`
- Add an internal session governance decorator that externalizes inline image/audio/file content to `artifact.Service` before persistence
- Keep runtime and persisted views separated: persisted events store `ContentRef`, while live session events retain original inline bytes for the current turn
- Hydrate `ContentRef` back to standard `ContentParts` on `GetSession`, full `ListSessions`, `SearchEvents`, and `GetEventWindow`
- Preserve optional session service capabilities (`SearchableService`, `WindowService`, `TrackService`) only when the wrapped backend originally implements them
- Add opt-in runner config via `runner.WithSessionMultimodalExternalization(...)` and document the feature in session docs
- Add end-to-end coverage for runner/session/artifact integration across in-memory, Redis, and optional MongoDB session backends
- Add a targeted runnable example for session multimodal externalization

### Model (`model/`)
- Add `ContentRef` field to `ContentPart`
- Add `ContentRef` metadata fields for artifact ref/name/version, MIME type, size, SHA256, original name, data URL source, and owner context

### Internal Session Governance (`internal/session/multimodal/`)
- Add session service decorator for multimodal externalization and hydration
- Externalize standard `ContentParts` inline bytes/data URLs for image/audio/file into artifact storage
- Store pinned refs as `artifact://<name>@<version>`
- Generate artifact names using `sessionpart_<unix-ms>_<sha256-16>_<uuid>.<ext>`
- Fail closed when artifact service is unavailable or artifact save/load fails
- Best-effort delete saved artifacts if append fails or if an append hook skips the actual write
- Skip partial/invalid events to avoid orphan artifacts for events that will not persist
- Preserve live-session semantics when persisted events are compacted, including `AppendEventHook` abort behavior
- Hydrate full session reads while preserving metadata-only list behavior
- Preserve inner optional interfaces and hydrate events returned by search/window APIs
- Add sentinel errors for artifact-service and artifact-ref failures so callers can use `errors.Is`
- Validate hydrated artifacts against `ContentRef` size/hash metadata and clear `ContentRef` after successful hydration

### Runner (`runner/`)
- Add `SessionMultimodalExternalizationConfig`
- Add `WithSessionMultimodalExternalization`
- Wrap session service during `NewRunner` and `NewRunnerWithAgentFactory` construction when enabled

### Docs and Examples (`docs/mkdocs/`, `examples/session/`)
- Document session multimodal externalization in English and Chinese session docs
- Cover opt-in usage, required artifact service, governance scope, hydrate behavior, skipped inputs, and fail-closed semantics
- Add `examples/session/multimodal_externalization`, a no-API-key example showing runtime inline content, persisted `ContentRef`, and second-turn hydrate behavior

### Tests
- Add tests for externalization and default hydration, including image/audio/file and data URL paths
- Add regression coverage for runtime session not being polluted by persisted events
- Add regression coverage for `AppendEventHook` skipping `next()`
- Add optional interface preservation coverage
- Add partial-event orphan prevention coverage
- Add `ListSessions`, `SearchEvents`, and `GetEventWindow` hydration coverage
- Add append-failure cleanup coverage
- Add artifact integrity validation and hydrated `ContentRef` clearing coverage
- Add `runner` option/wrapper assembly coverage, including `NewRunnerWithAgentFactory`
- Add `test/` end-to-end coverage for runner + session backend + artifact service + model request boundary
- Cover in-memory and Redis session backends in E2E; MongoDB E2E runs when `MONGO_TEST_URI` is provided

## Test plan

- [x] `GOPROXY=https://proxy.golang.org,direct go test ./internal/session/multimodal ./model ./runner`
- [x] `go test ./internal/session/multimodal ./runner`
- [x] `cd test && go test ./...`
- [x] `cd test && go test ./... -run 'TestSessionMultimodal'`
- [x] `cd examples/session && GOPROXY=https://proxy.golang.org,direct go run ./multimodal_externalization`
- [x] `cd examples/session && GOPROXY=https://proxy.golang.org,direct go test ./multimodal_externalization`
- [x] Linux remote full test: `go test ./...`
  - Go: `/tmp/go-toolchains/go1.26.4/bin/go`
  - Python: `python3.8` on `PATH`
  - Run as non-root user
  - `TMPDIR=/tmp`
- [x] `ReadLints` on changed files: no diagnostics

### CODE REVIEW

File: internal/session/externalization/service.go [hunk 1]
```diff
@@ -0,0 +1,1076 @@
...
+	}
+	if !changed {
+		return s.Service.AppendEvent(ctx, sess, evt, options...)
+	}
+	persistedSess := sess.Clone()
+	if err := s.Service.AppendEvent(ctx, persistedSess, persisted, options...); err != nil {
+		bestEffortDelete(ctx, s.artifactService, info, saved)
+		return err
+	}
+	sess.UpdateUserSession(evt, options...)
```
Comment (Flash-LHR) [inline internal/session/externalization/service.go:275]: This live-session update runs after the inner hook chain, so an `AppendEventHook` that skips `next` still leaves the event in memory. Move this update into the hook-guarded append path.
<details><summary>中文</summary>
这个实时会话更新在内部 hook 链之后执行，所以跳过 `next` 的 `AppendEventHook` 仍会把事件留在内存里。请把这个更新移到受 hook 保护的追加路径中。
</details>

Comment (hr-chang) [reply internal/session/externalization/service.go:275]: 已修复。

之前外层 wrapper 在 inner `AppendEvent` 返回 nil 后会无条件更新 live session；如果底层 `AppendEventHook` 通过不调用 `next()` 跳过实际 append，就会出现 persisted 层未写入、但 live session 被更新的问题。
现在改为：外存路径调用 inner append 后，会先确认 hook-guarded append 确实新增了本次 event；如果 hook 跳过 `next()`，wrapper 会清理本次已保存的 artifact，并保持 live session 不变。已补充回归测试 `TestAppendEventHookSkipNextDoesNotUpdateLiveSession` 覆盖该场景。

File: runner/runner.go [hunk 2]
```diff
@@ -362,6 +383,18 @@ func newOptions(opt ...Option) Options {
 	return opts
 }
 
+func wrapSessionMultimodalExternalization(options Options) session.Service {
+	cfg := options.sessionMultimodalExternalization
+	if !cfg.Enabled {
+		return options.sessionService
```
Comment (hyprh) [inline runner/runner.go]: 之类如果artifact service 为 nil 是否要拦截一下

Comment (hr-chang) [reply runner/runner.go]: 我判断不应拦截，会增加业务方排查难度。
当前artifactService==nil会在各实际调用artifact阶段fail，业务方在研发阶段一定能快速感知到；但自动降级的话，这就成暗坑了。此外，在实际调用时fail的编码风格与仓库类似场景风格一致。

Comment (hyprh) [reply runner/runner.go]: 感觉应该可以先给一个ERROR 日志？

File: ..dev/.checklist/code.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+
+# 代码规范
+
+- 导出兼容性: 除非有充分且明确的理由，否则不应改变已有导出项的兼容性。
+    > 目标是让业务方升级框架版本时，不因兼容性变化引入额外问题。
```
Comment (hyprh) [inline ..dev/.checklist/code.md]: 这个文件应该不需要合入吧

Comment (hr-chang) [reply ..dev/.checklist/code.md]: 诶，这是个意外，我移除下。

File: examples/session/externalization/main.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+			runner.SessionMultimodalExternalizationConfig{Enabled: true},
+		),
+	)
+	defer func() {
+		if err := r.Close(); err != nil {
+			log.Printf("close runner: %v", err)
+		}
+	}()
+
+	imageData := []byte("tiny-demo-image")
```
Comment (hyprh) [inline examples/session/externalization/main.go:77]: 示例可以用日志展示一下实际存储的内容，以及发给LLM的内容 分别是怎么样的

File: runner/runner.go (hunk 2 above)
Comment (hr-chang) [reply runner/runner.go]: 新版入口已改为显式 session/multimodal.Wrap(sessionSvc, artifactSvc, cfg)，artifact service 在调用签名里就是启用该能力的必要依赖。这里不额外打构造期 ERROR 日志；若误传 nil，实际 save/load/hydrate 时会 fail closed 返回明确的 ErrArtifactServiceNil，避免静默降级或 panic。

File: examples/session/externalization/main.go (hunk 4 above)
Comment (hr-chang) [reply examples/session/externalization/main.go:77]: 已调整 example，现在输出会分开展示：

发给 LLM 的 request 视图：展示 image bytes 长度、file data URL 是否存在、以及是否包含内部 ContentRef。
实际持久化到 session backend 的 event 视图：展示 image bytes 长度、file URL 长度，以及 image/file 对应的 artifact://... 引用。
同时补充了第二轮同 session 的 user turn，用来展示历史 session 中的 ContentRef 会在构造下一次模型请求前 hydrate 回正常 ContentParts。README 里也补充了业务接入步骤和对应输出示例。

**Review by hyprh — COMMENTED** [review]
Two blocking issues inline (both are "delete an artifact that a persisted event still references" hazards — with fail-closed hydration that permanently breaks reads for the session), plus one compat note:

**Externalization is effectively one-way.** Once sessions contain `ContentRef`s, removing the wrapper (config rollback) or reading with an older library version (unknown `content_ref` dropped on unmarshal) yields image parts with nil `Data`: the openai adapter then silently sends an empty `data:image/png;base64,` (anthropic/bedrock at least error out), and the openai batch path marshals `content_ref` straight to the provider wire. The docs should call out that the wrapper must not be removed while externalized sessions are still being read.

File: internal/session/externalization/service.go [hunk 5]
```diff
@@ -0,0 +1,1110 @@
...
+	if !changed {
+		return s.Service.AppendEvent(ctx, sess, evt, options...)
+	}
+	persistedSess := sess.Clone()
+	beforeEvents := len(persistedSess.Events)
+	if err := s.Service.AppendEvent(ctx, persistedSess, persisted, options...); err != nil {
+		bestEffortDelete(ctx, s.artifactService, info, saved)
+		return err
+	}
+	if !appendObserved(persistedSess, persisted.ID, beforeEvents) {
```
Comment (hyprh) [inline internal/session/externalization/service.go:272]: `appendObserved` uses "event visible in the cloned session" as the persisted-oracle, but that contract doesn't hold: every backend's `AppendEvent` starts with `sess.UpdateUserSession(e, opts...)`, whose `ApplyEventFiltering` clears `sess.Events` when the session contains no user message — while the storage layer still persists the event.

So when the first persisted event of a session is a non-user event with externalizable content (e.g. seed history starting with system/assistant events, or direct `AppendEvent` calls), the flow becomes: artifact saved → event persisted with `ContentRef` → event ID not found in the cleared clone → `bestEffortDelete` removes the artifacts just saved. The session is left with dangling refs, and `GetEventWindow`/`SearchEvents` fail permanently with `artifact not found` (reproduced on inmemory; the window/search read paths don't apply user-anchored trimming, so they do surface the event).

This needs a persistence signal that doesn't rely on the runtime view of the passed-in session.

File: internal/session/externalization/service.go [hunk 6]
```diff
@@ -0,0 +1,1110 @@
...
+		bestEffortDelete(ctx, s.artifactService, info, saved)
+		return err
+	}
+	if !changed {
+		return s.Service.AppendEvent(ctx, sess, evt, options...)
+	}
+	persistedSess := sess.Clone()
+	beforeEvents := len(persistedSess.Events)
+	if err := s.Service.AppendEvent(ctx, persistedSess, persisted, options...); err != nil {
+		bestEffortDelete(ctx, s.artifactService, info, saved)
```
Comment (hyprh) [inline internal/session/externalization/service.go]: Same deletion hazard on this error path: an append hook may call `next()` (event persisted) and then return an error. This branch can't distinguish "not persisted" from "persisted, but a hook failed afterwards", yet it deletes artifacts the persisted event now references — with fail-closed hydration the whole session becomes unreadable.

Deleting a referenced artifact is much worse than leaking an orphan; cleanup should only run when the event is known not to have reached storage.

Comment (WineChord) [inline internal/session/externalization/service.go]: If an `AppendEventHook` calls `next()` and then returns an error, the event may already be persisted with `ContentRef`, but this error branch deletes the saved artifacts anyway. The persisted event then points to deleted artifacts and later hydration fails. Please only cleanup when append is known not to have reached storage; if the result is ambiguous, leaving orphan artifacts is safer than deleting referenced artifacts.

<details><summary>中文</summary>
如果 `AppendEventHook` 先调用了 `next()`，随后返回 error，事件可能已经带着 `ContentRef` 持久化了，但这里的错误分支仍会删除刚保存的 artifacts。这样持久化事件会指向已删除的 artifact，后续 hydrate 会失败。建议只在确定 append 没有进入存储时 cleanup；如果无法确定，保留孤儿 artifact 比删除仍被引用的 artifact 更安全。
</details>

File: internal/session/externalization/service.go (hunk 5 above)
Comment (WineChord) [inline internal/session/externalization/service.go:272]: `appendObserved` uses the cloned runtime session's visible `Events` as the persistence signal, but `UpdateUserSession` can filter that view to empty when there is no user message even though the backend persisted the event. First system/assistant/direct appended multimodal events can therefore persist dangling `ContentRef`s and break `GetEventWindow` / `SearchEvents` with artifact-not-found. Please replace this with a real persisted/next-called signal, or stop deleting artifacts on this ambiguous path; add regression coverage for a first non-user externalized event.

<details><summary>中文</summary>
`appendObserved` 用 cloned runtime session 里可见的 `Events` 判断是否持久化，但当没有 user message 时，`UpdateUserSession` 可能把这个视图过滤为空，即使后端已经写入了事件。这样首个 system/assistant/direct append 的 multimodal event 可能留下悬空 `ContentRef`，导致 `GetEventWindow` / `SearchEvents` hydrate 时 artifact-not-found。建议改成真实的 persisted/next-called 信号，或者在这个歧义路径上不要删除 artifacts；并补一个首个非 user externalized event 的回归测试。
</details>

**Review by WineChord — CHANGES_REQUESTED** [review]
I found a remaining cleanup/persistence ambiguity that can leave persisted events pointing to deleted artifacts.

File: internal/session/externalization/service.go (hunk 5 above)
Comment (hr-chang) [reply internal/session/externalization/service.go:272]: 感谢指出，这里确实不能把 cloned runtime session view 当作 persisted oracle。

已调整：`appendObserved` 现在只用于决定是否同步 caller-owned live session，不再触发 artifact 删除。进入 inner `AppendEvent` 之后，如果 cloned session 因 filtering / user-message anchoring 看不到 event，也会保留 artifacts，避免删除已被持久化 event 引用的内容。

已补充 `TestAppendEventKeepsArtifactsWhenFilteredFromRuntimeView`，覆盖首个非 user externalized event 的场景，并验证 `GetEventWindow` 可以正常 hydrate。

File: internal/session/externalization/service.go (hunk 6 above)
Comment (hr-chang) [reply internal/session/externalization/service.go]: 同意，这个 error path 也属于 append 结果不确定场景。

已调整：一旦 event 交给 inner `AppendEvent`，返回 error 时不再删除本次 artifacts。这样即使 hook 已经调用 `next()` 完成持久化后再返回 error，也不会留下 dangling `ContentRef`。这里选择保留可能的 orphan artifact，因为它比删除已被持久化 event 引用的 artifact 风险更低。

已补充 `TestAppendEventHookErrorAfterNextKeepsReferencedArtifacts` 覆盖该场景。

另外，文档也已补充 externalization 的持续读取约束：只要已外存 session 仍可能被读取，就需要继续使用 wrapped session service；移除 wrapper 或使用旧版本读取前，需要先迁移、清理或等待相关 session/artifact 过期。

File: internal/session/externalization/service.go (hunk 5 above)
Comment (hr-chang) [reply internal/session/externalization/service.go:272]: 感谢，这条与 hyprh 指出的 `appendObserved` 不能作为 persistence signal 的问题一致，判断成立，已由同一组修复覆盖。

`appendObserved` 现在只控制 caller-owned live session 同步，不再触发 artifact 删除；进入 inner `AppendEvent` 后，即使 runtime view 因 filtering 看不到 event，也会保留 artifacts。已补充 `TestAppendEventKeepsArtifactsWhenFilteredFromRuntimeView`。

File: model/request.go [hunk 7]
```diff
@@ -270,6 +270,38 @@ type ContentPart struct {
...
+	SHA256 string `json:"sha256,omitempty"`
+	// OriginalName is the user-facing filename or display name, if any.
+	OriginalName string `json:"original_name,omitempty"`
+	// FromDataURL indicates whether the content was externalized from a data URL.
+	FromDataURL bool `json:"from_data_url,omitempty"`
+	// EventID identifies the owning event for debugging.
+	EventID string `json:"event_id,omitempty"`
+	// ChoiceIndex identifies the owning response choice within the event.
+	ChoiceIndex int `json:"choice_index"`
+	// PartIndex identifies the owning content part within the message.
```
Comment (hyprh) [inline model/request.go]: 这里index的字段还有fromdataurl 感觉是不是不是必要的

Comment (hr-chang) [reply model/request.go]: 确实。
这几个字段首期不参与 hydrate / artifact lookup / restore 逻辑，更多是 debug metadata。考虑到 `ContentRef` 会持久化到 session backend，schema 应尽量收敛，避免写入不必要字段。

已移除 `FromDataURL`、`ChoiceIndex`、`PartIndex`，并同步清理赋值和测试断言。

**Review by WineChord — APPROVED** [review]
LGTM
